lib/vector-core/src/event/value.rs

```diff
@@ -22,6 +23,41 @@ pub enum Value {
     Null,
 }
 
+impl Eq for Value {}
+
+impl Hash for Value {
```

Comment:
I think there is a risk here that two different enum variants could hash to the same value if their values hash to the same value. When Rust derives hashes on enums, it appears to first hash `core::mem::discriminant` so that each enum hashes to a different value even if the values wrapped by the enum variant would have hashed to the same thing.

That is:

```
Value::Array(v) => {
    v.hash(state);
}
```

Would go to:

```
Value::Array(v) => {
  core::mem::discriminant(self).hash(state);
  v.hash(state);
}
```

You could probably pull up `core::mem::discriminant(self).hash(state)` to the top since each enum variant hash implementation would do it.